benchmark/multi_gpu/training/training_benchmark.py

```diff
@@ -0,0 +1,300 @@
+        err_msg = (f'Configuration of {args.dataset} + {args.model}'
+                   'not supported')
+        raise RuntimeError(err_msg)
+    if rank == 0:
+        print(f'Training bench for {args.model}:')
+
+    num_nodes = int(mask[-1].sum()) if hetero else int(mask.sum())
+    num_neighbors = args.num_neighbors
+
+    if type(num_neighbors) is list:
+        if len(num_neighbors) == 1:
+            num_neighbors = num_neighbors * args.num_layers
+    elif type(num_neighbors) is int:
+        num_neighbors = [num_neighbors] * args.num_layers
+
+    if len(num_neighbors) != args.num_layers:
+        err_msg = (f'num_neighbors={num_neighbors} lenght != num of'
+                   'layers={args.num_layers}')
+
+    kwargs = {
+        'num_neighbors': num_neighbors,
+        'batch_size': args.batch_size,
+        'num_workers': args.num_workers,
+    }
+    subgraph_loader = NeighborLoader(
+        data,
+        input_nodes=mask,
+        sampler=None,
+        **kwargs,
+    )
+    if rank == 0 and args.evaluate:
+        val_loader = NeighborLoader(
+            data,
+            input_nodes=val_mask,
+            sampler=None,
+            **kwargs,
+        )
+        test_loader = NeighborLoader(
+            data,
+            input_nodes=test_mask,
+            sampler=None,
+            **kwargs,
+        )
+
+    if rank == 0:
+        print('----------------------------------------------')
+        print(
+            f'Batch size={args.batch_size}, '
+            f'Layers amount={args.num_layers}, '
+            f'Num_neighbors={num_neighbors}, '
+            f'Hidden features size={args.num_hidden_channels}', flush=True)
+
+    params = {
+        'inputs_channels': inputs_channels,
+        'hidden_channels': args.num_hidden_channels,
+        'output_channels': num_classes,
+        'num_heads': args.num_heads,
+        'num_layers': args.num_layers,
+    }
+
+    if args.model == 'pna' and degree is None:
+        degree = PNAConv.get_degree_histogram(subgraph_loader)
+        print(f'Rank: {rank}, calculated degree for {args.dataset}.',
+              flush=True)
+        params['degree'] = degree
+    dist.barrier()
+
+    torch.manual_seed(12345)
+    model = get_model(args.model, params,
+                      metadata=data.metadata() if hetero else None)
+    model = model.to(device)
+    if hetero:
+        # TODO (DamianSzwichtenberg):
+        # Provide fix for:
+        # RuntimeError: Modules with uninitialized parameters can't be used
+        # with `DistributedDataParallel`. Run a dummy forward pass to correctly
+        # initialize the modules.
+        pass
+    model = DDP(model, device_ids=[device])
+    model.train()
+
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
+
+    if args.device == 'xpu':
+        model, optimizer = ipex.optimize(model, optimizer=optimizer)
+
+    train = train_hetero if hetero else train_homo
+
+    if rank == 0:
+        maybe_synchronize(args.device)
```

Comment:
n00b question: Is it correct that it's synchronised on rank zero only? Should it be on all ranks?

Reply:
That's a very good question and a good catch. Indeed I should synchronize on all ranks. Thanks 😉 just fixed that.